### PR TITLE
[Console] Document console cursor

### LINKED ISSUE #13529: [Console] Add Cursor class to control the cursor in the terminal
| Q            | A
| ------------ | ---
| Feature PR   | symfony/symfony#27444
| PR author(s) | @pierredup
| Merged in    | 5.1-dev

### PR DESCRIPTION
**Pull request by noniagriconomie**
fixes https://github.com/symfony/symfony-docs/issues/13529
feature https://github.com/symfony/symfony/pull/27444

ping @wouterj I've seen your slack recently and wanted to give it a try as I will use the feature very soon :)
feel free to comment, specialy if I need to document all the cursor features or not

text and image taken from the feature PR description
inspired by https://symfony.com/doc/current/components/console/helpers/progressbar.html

### CODE REVIEW

**Comment by Nyholm** [comment]
(The “code block” ci job does not support adding new files. I’ll fix that in a separate pr)

**Comment by Nyholm** [comment]
The CI job is updated. 

Regarding your PR: Do we need to create a new page or is there an existing page we can add this content to?

**Comment by noniagriconomie** [comment]
@Nyholm Thank you for the CI
for your second point, I personally checked the https://symfony.com/doc/current/components/console.html#learn-more list and found it will be great to add it here

If you or other found a better place,, I will be pleased to swap the code of course :)

**Comment by Nyholm** [comment]
Oh. Hmm.. I see. 

There is a greater plan to make everything under one document. I'll wait for input from the docs team. 

--------

Btw, test retrigger the CI with:

```
git commit --allow-empty  -m "Trigger CI"
git push
```

**Comment by Nyholm** [comment]
Cool. Another CI error. We dont support non-rst files =) 

I should filter out these too.

**Comment by Nyholm** [comment]
May I bother you to trigger the CI again?

**Comment by noniagriconomie** [comment]
@Nyholm here you are :)
feel free to edit/commit on this PR if you can

**Comment by Nyholm** [comment]
Wohoo. The CI is green. Thank you for helping me. 

------

I am also happy with the content of this PR. 
👍🏽

File: components/console/helpers/cursor.rst [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+Cursor Helper
+=============
+
+.. versionadded:: 5.1
+
+    The :class:`Symfony\\Component\\Console\\Cursor`
+    class was introduced in Symfony 5.1.
+
+The :class:`Symfony\\Component\\Console\\Cursor` allows you to manipulate
+the cursor in a console command like this:
```
Comment (wouterj) [inline components/console/helpers/cursor.rst]:
```suggestion
The :class:`Symfony\\Component\\Console\\Cursor` allows you to change the
cursor position in a console command. This allows you to write on any position
of the output:
```

File: components/console/helpers/cursor.rst [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+.. image:: /_images/components/console/cursor.gif
+   :align: center
+
```
Comment (wouterj) [inline components/console/helpers/cursor.rst:18]: Then, I would add a small code example after this gif and especially emphasize that the Cursor only moves the position and you should use the output to write on that position. E.g.

```php
// ...
use Symfony\Component\Console\Cursor;

class AcmeCommand extends Command
{
    // ...

    protected function execute(InputInterface $input, OutputInterface $output): int
    {
        // ...

        $cursor = new Cursor($output);

        // move to a specific column and row using the cursor
        $cursor->moveToPosition(7, 11);
        // and write text on this position using the output
        $output->write('Some text');
    }
}
```

File: components/console/helpers/cursor.rst [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    // moves the cursor to column and row
+    $cursor->moveToPosition(7, 11);
+
+    // moves the cursor to the right
+    $cursor->moveRight();
+
+    // clear the current cursor line
+    $cursor->clearLine();
+
+    // [...] read the class to see all the features
```
Comment (wouterj) [inline components/console/helpers/cursor.rst]: I'm in favor of documenting all features in 2 small sections/code examples:

1. Moving the cursor, explaining all `Cursor::move*()` methods. You can show multiple methods with 1 comment, e.g. "Move the cursor to the right/left, or up/down from its current position" and then show moveRight, moveLeft, moveUp, moveDown)
2. Clearing output, explainig all `Cursor::clear*()` methods (these are especially a bit vague, e.g. `clearScreen()` only clears from the current position to the end)

And then show `$cursor->hide()` and `->show()` somewhere as well :)

**Review by wouterj — CHANGES_REQUESTED** [review]
Thanks a lot for starting this @noniagriconomie!

I think we should document all available methods. I've made some suggestions on how to do this somewhat briefly below (to avoid long boring listings).

I'm fine with keeping it in a separate document. It seems the most logical from the current Console documentation structure. We can always revisit this when we start working on the Console docs

**Comment by noniagriconomie** [comment]
@wouterj many thanks for the reviews, I've tried to address them all
status: review

edit: I do not understand the [CI sphinx error](https://github.com/symfony/symfony-docs/runs/2445226351?check_suite_focus=true)

File: components/console/helpers/cursor.rst [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    class was introduced in Symfony 5.1.
+
+The :class:`Symfony\\Component\\Console\\Cursor` allows you to change the
+cursor position in a console command. This allows you to write on any position
+of the output:
+
+.. image:: /_images/components/console/cursor.gif
+   :align: center
+
+
```
Comment (wouterj) [inline components/console/helpers/cursor.rst:19]:
```suggestion
.. code-block:: php

```

This should fix the Sphinx error :)

File: components/console/helpers/cursor.rst [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    $cursor->moveRight();
+
+    // moves the cursor 3 columns right from its current position
+    $cursor->moveRight(3);
+
+    // same for left
+    $cursor->moveLeft();
+
+    // move the cursor to a specific position from its current position
+    $cursor->moveToPosition(7, 11);
```
Comment (wouterj) [inline components/console/helpers/cursor.rst:75]: I haven't used this feature, but I think we have to clearify this a bit. This means 7 lines down, and 11 columns to the right? And I can use negative numbers to achieve the opposite?

Or does this expect a coordinate?

**Review by wouterj — APPROVED** [review]
Thanks! 2 minor comments left, ready to merge imho :)

File: components/console/helpers/cursor.rst (hunk 5 above)
Comment (noniagriconomie) [reply components/console/helpers/cursor.rst:75]: @wouterj it means 7 cols right, 11 rows down from console init cursor
the cursor always moves position from this 0:0 position, thus need an x:y (here positive values) from the init yes

**Comment by noniagriconomie** [comment]
friendly ping @OskarStark I may hope you are the best to ping here related to the error I have:

```
Warning: ] Found "1" invalid file!                                              
Error: Please do not use ".. code-block:: php", use "::" instead.
```

I've check other code bloc and for me it is "correct" here, or what am I doing wrong? many thanks :)

**Comment by wouterj** [comment]
Thank you @noniagriconomie ! I'm sorry this took a while...

The build error appears to be a bug in DOCtor (@OskarStark), I've silenced the rule for now (along with some minor changes) in 7565926922b3e680c1b33664a6bd928c7d666206

**Pull request merged**